File: Financial Fraud Detection using Predictive Analytics/scripts/r/ml_analysis.R
# Machine Learning Analysis Script for Fraud Detection (R)
#
# This script implements machine learning models for fraud detection:
# 1. Data preprocessing and feature engineering
# 2. Model training (Logistic Regression, Random Forest, XGBoost)
# 3. Model evaluation and validation
# 4. Feature importance analysis

# Load libraries
library(tidyverse)
library(caret)
library(randomForest)
library(xgboost)
library(pROC)
library(ROCR)
library(ggplot2)

# Set options
options(warn = -1)
set.seed(42)

# Set up paths
project_root <- dirname(dirname(dirname(getwd())))
data_path <- file.path(project_root, "data", "fraud_data.csv")
output_dir <- file.path(project_root, "outputs", "figures")
model_dir <- file.path(project_root, "outputs", "models")

# Create output directories if they don't exist
dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
dir.create(model_dir, recursive = TRUE, showWarnings = FALSE)

# Load data
cat("Loading data...\n")
df <- read.csv(data_path, stringsAsFactors = FALSE)
cat("Data loaded:", dim(df), "\n")
cat("Fraud rate:", mean(df$isFraud), "\n")
cat("Target distribution:\n")
print(table(df$isFraud))

# ============================================================================
# 1. DATA PREPROCESSING
# ============================================================================
cat("\n", rep("=", 80), "\n", sep = "")
cat("DATA PREPROCESSING\n")
cat(rep("=", 80), "\n", sep = "")

# Select features for modeling
key_features <- c("TransactionAmt", "card1", "card2", "card3", "card5", 
                  "addr1", "addr2", "dist1", "dist2")
key_features <- key_features[key_features %in% colnames(df)]

# Add some C and D features if available
c_features <- grep("^C[0-9]+$", colnames(df), value = TRUE)[1:10]
d_features <- grep("^D[0-9]+$", colnames(df), value = TRUE)[1:10]

features <- c(key_features, c_features, d_features)
features <- features[features %in% colnames(df)]

cat("Selected", length(features), "features for modeling\n")

# Prepare data
X <- df[features]
y <- df$isFraud

# Handle missing values
for(col in features) {
  if(any(is.na(X[[col]]))) {
    X[[col]][is.na(X[[col]])] <- median(X[[col]], na.rm = TRUE)
  }
}

cat("X shape:", dim(X), "\n")
cat("y shape:", length(y), "\n")
cat("Missing values in X:", sum(is.na(X)), "\n")

# Split data into train and test sets
trainIndex <- createDataPartition(y, p = 0.8, list = FALSE)
X_train <- X[trainIndex, ]
X_test <- X[-trainIndex, ]
y_train <- y[trainIndex]
y_test <- y[-trainIndex]

cat("\nTrain set:", dim(X_train), "Fraud rate:", mean(y_train), "\n")
cat("Test set:", dim(X_test), "Fraud rate:", mean(y_test), "\n")

# ============================================================================
# 2. MODEL TRAINING
# ============================================================================
cat("\n", rep("=", 80), "\n", sep = "")
cat("MODEL TRAINING\n")
cat(rep("=", 80), "\n", sep = "")

models <- list()
predictions <- list()
predictions_proba <- list()

# Model 1: Logistic Regression
cat("\nTraining Logistic Regression...\n")
train_data <- cbind(X_train, isFraud = y_train)
lr_model <- glm(isFraud ~ ., data = train_data, family = binomial)
lr_pred_proba <- predict(lr_model, newdata = X_test, type = "response")
lr_pred <- ifelse(lr_pred_proba > 0.5, 1, 0)

models[["Logistic Regression"]] <- lr_model
predictions[["Logistic Regression"]] <- lr_pred
predictions_proba[["Logistic Regression"]] <- lr_pred_proba

cat("Logistic Regression trained!\n")
cat("AUC-ROC:", auc(roc(y_test, lr_pred_proba)), "\n")
cat("Accuracy:", mean(lr_pred == y_test), "\n")

# Model 2: Random Forest
cat("\nTraining Random Forest...\n")
train_data <- cbind(X_train, isFraud = as.factor(y_train))
rf_model <- randomForest(isFraud ~ ., data = train_data,
                         ntree = 100, mtry = sqrt(ncol(X_train)),
                         classwt = c(1, sum(y_train == 0) / sum(y_train == 1)))
rf_pred_proba <- predict(rf_model, newdata = X_test, type = "prob")[, 2]
rf_pred <- predict(rf_model, newdata = X_test)

models[["Random Forest"]] <- rf_model
predictions[["Random Forest"]] <- as.numeric(rf_pred) - 1
predictions_proba[["Random Forest"]] <- rf_pred_proba

cat("Random Forest trained!\n")
cat("AUC-ROC:", auc(roc(y_test, rf_pred_proba)), "\n")
cat("Accuracy:", mean((as.numeric(rf_pred) - 1) == y_test), "\n")

# Model 3: XGBoost
cat("\nTraining XGBoost...\n")
dtrain <- xgb.DMatrix(data = as.matrix(X_train), label = y_train)
dtest <- xgb.DMatrix(data = as.matrix(X_test), label = y_test)

xgb_model <- xgboost(data = dtrain, nrounds = 100, max_depth = 6, 
                     eta = 0.1, objective = "binary:logistic", 
                     eval_metric = "auc", verbose = 0,
                     scale_pos_weight = sum(y_train == 0) / sum(y_train == 1))
xgb_pred_proba <- predict(xgb_model, dtest)
xgb_pred <- ifelse(xgb_pred_proba > 0.5, 1, 0)

models[["XGBoost"]] <- xgb_model
predictions[["XGBoost"]] <- xgb_pred
predictions_proba[["XGBoost"]] <- xgb_pred_proba

cat("XGBoost trained!\n")
cat("AUC-ROC:", auc(roc(y_test, xgb_pred_proba)), "\n")
cat("Accuracy:", mean(xgb_pred == y_test), "\n")

# ============================================================================
# 3. MODEL EVALUATION
# ============================================================================
cat("\n", rep("=", 80), "\n", sep = "")
cat("MODEL EVALUATION\n")
cat(rep("=", 80), "\n", sep = "")

# ROC curves
roc_curves <- lapply(predictions_proba, function(pred) roc(y_test, pred))

# Plot ROC curves
png(file.path(output_dir, "roc_curves_r.png"), width = 1200, height = 1000, res = 300)
plot(roc_curves[[1]], main = "ROC Curve Comparison", col = 2, lwd = 2)
for(i in 2:length(roc_curves)) {
  lines(roc_curves[[i]], col = i + 1, lwd = 2)
}
legend("bottomright", legend = names(roc_curves), 
       col = 2:(length(roc_curves) + 1), lwd = 2)
abline(a = 0, b = 1, lty = 2, col = "gray")
dev.off()
cat("\nROC curves saved to", file.path(output_dir, "roc_curves_r.png"), "\n")

# Model comparison
comparison <- data.frame(
  Model = names(predictions_proba),
  AUC_ROC = sapply(roc_curves, function(r) as.numeric(auc(r))),
  Accuracy = sapply(1:length(predictions), function(i) {
    mean(predictions[[i]] == y_test)
  })
)

cat("\nModel Comparison:\n")
cat(rep("=", 80), "\n", sep = "")
print(comparison[order(comparison$AUC_ROC, decreasing = TRUE), ])

# ============================================================================
# 4. FEATURE IMPORTANCE ANALYSIS
# ============================================================================
cat("\n", rep("=", 80), "\n", sep = "")
cat("FEATURE IMPORTANCE ANALYSIS\n")
cat(rep("=", 80), "\n", sep = "")

# Random Forest feature importance
if("Random Forest" %in% names(models)) {
  rf_importance <- data.frame(
    feature = features,
    importance = importance(models[["Random Forest"]])
  )
  rf_importance <- rf_importance[order(rf_importance$importance, decreasing = TRUE), ]
  rf_importance_top <- head(rf_importance, 20)
  
  png(file.path(output_dir, "feature_importance_rf_r.png"), 
      width = 1200, height = 800, res = 300)
  barplot(rf_importance_top$importance, names.arg = rf_importance_top$feature,
          main = "Random Forest - Top 20 Feature Importance",
          xlab = "Importance", ylab = "Feature", horiz = TRUE, las = 1)
  dev.off()
  cat("\nFeature importance plot saved to", 
      file.path(output_dir, "feature_importance_rf_r.png"), "\n")
  
  cat("\nTop 10 Features (Random Forest):\n")
  print(head(rf_importance, 10))
}

# XGBoost feature importance
if("XGBoost" %in% names(models)) {
  xgb_importance <- xgb.importance(feature_names = features, model = models[["XGBoost"]])
  xgb_importance_top <- head(xgb_importance, 20)
  
  png(file.path(output_dir, "feature_importance_xgb_r.png"), 
      width = 1200, height = 800, res = 300)
  xgb.plot.importance(xgb_importance_top, main = "XGBoost - Top 20 Feature Importance")
  dev.off()
  cat("\nFeature importance plot saved to", 
      file.path(output_dir, "feature_importance_xgb_r.png"), "\n")
  
  cat("\nTop 10 Features (XGBoost):\n")
  print(head(xgb_importance, 10))
}

# ============================================================================
# 5. SAVE BEST MODEL
# ============================================================================
cat("\n", rep("=", 80), "\n", sep = "")
cat("SAVING BEST MODEL\n")
cat(rep("=", 80), "\n", sep = "")

# Find best model
best_model_name <- comparison$Model[which.max(comparison$AUC_ROC)]
best_model <- models[[best_model_name]]

# Save model
saveRDS(best_model, file.path(model_dir, "best_model_r.rds"))
cat("\nBest Model:", best_model_name, "\n")
cat("Best AUC-ROC:", max(comparison$AUC_ROC), "\n")
cat("Model saved to", file.path(model_dir, "best_model_r.rds"), "\n")

# Summary
cat("\n", rep("=", 80), "\n", sep = "")
cat("MACHINE LEARNING ANALYSIS SUMMARY\n")
cat(rep("=", 80), "\n", sep = "")
cat("\nBest Model:", best_model_name, "\n")
cat("Best AUC-ROC:", max(comparison$AUC_ROC), "\n")
cat("\nModels trained:", length(models), "\n")
cat("Features used:", length(features), "\n")
cat("\nAnalysis complete!\n")
cat(rep("=", 80), "\n", sep = "")

File: Financial Fraud Detection using Predictive Analytics/scripts/python/ml_analysis.py
"""
Machine Learning Analysis Script for Fraud Detection

This script implements machine learning models for fraud detection:
1. Data preprocessing and feature engineering
2. Model training (Logistic Regression, Random Forest, XGBoost, LightGBM)
3. Model evaluation and validation
4. Feature importance analysis
5. Model interpretation
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path
import warnings
from sklearn.model_selection import train_test_split, cross_val_score, StratifiedKFold
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score, roc_curve, precision_recall_curve
import xgboost as xgb
import lightgbm as lgb
from imblearn.over_sampling import SMOTE
import joblib

warnings.filterwarnings('ignore')
sns.set_style("whitegrid")
np.random.seed(42)

def load_and_prepare_data(data_path):
    """Load and prepare data for modeling."""
    print("Loading data...")
    df = pd.read_csv(data_path)
    print(f"Data loaded: {df.shape}")
    print(f"Fraud rate: {df['isFraud'].mean():.4f}")
    print(f"Target distribution:\n{df['isFraud'].value_counts()}")
    
    # Select features for modeling
    key_features = ['TransactionAmt', 'card1', 'card2', 'card3', 'card5', 
                    'addr1', 'addr2', 'dist1', 'dist2']
    key_features = [f for f in key_features if f in df.columns]
    
    # Add some C and D features if available
    c_features = [col for col in df.columns if col.startswith('C') and col[1:].isdigit()][:10]
    d_features = [col for col in df.columns if col.startswith('D') and col[1:].isdigit()][:10]
    
    features = key_features + c_features + d_features
    features = [f for f in features if f in df.columns]
    
    print(f"\nSelected {len(features)} features for modeling")
    
    # Prepare data
    X = df[features].copy()
    y = df['isFraud'].copy()
    
    # Handle missing values
    X = X.fillna(X.median())
    
    # Handle infinite values
    X = X.replace([np.inf, -np.inf], np.nan)
    X = X.fillna(X.median())
    
    print(f"X shape: {X.shape}")
    print(f"y shape: {y.shape}")
    print(f"Missing values in X: {X.isnull().sum().sum()}")
    
    return X, y, features

def train_models(X_train, X_test, y_train, y_test, features, output_dir):
    """Train multiple ML models and evaluate them."""
    print("\n" + "="*80)
    print("MODEL TRAINING")
    print("="*80)
    
    models = {}
    predictions = {}
    predictions_proba = {}
    
    # Scale features for logistic regression
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)
    
    # Model 1: Logistic Regression
    print("\nTraining Logistic Regression...")
    lr_model = LogisticRegression(random_state=42, max_iter=1000, class_weight='balanced')
    lr_model.fit(X_train_scaled, y_train)
    lr_pred = lr_model.predict(X_test_scaled)
    lr_pred_proba = lr_model.predict_proba(X_test_scaled)[:, 1]
    
    models['Logistic Regression'] = lr_model
    predictions['Logistic Regression'] = lr_pred
    predictions_proba['Logistic Regression'] = lr_pred_proba
    
    print("Logistic Regression trained!")
    print(f"AUC-ROC: {roc_auc_score(y_test, lr_pred_proba):.4f}")
    print(f"Accuracy: {(lr_pred == y_test).mean():.4f}")
    
    # Model 2: Random Forest
    print("\nTraining Random Forest...")
    rf_model = RandomForestClassifier(
        n_estimators=100,
        max_depth=10,
        random_state=42,
        class_weight='balanced',
        n_jobs=-1
    )
    rf_model.fit(X_train, y_train)
    rf_pred = rf_model.predict(X_test)
    rf_pred_proba = rf_model.predict_proba(X_test)[:, 1]
    
    models['Random Forest'] = rf_model
    predictions['Random Forest'] = rf_pred
    predictions_proba['Random Forest'] = rf_pred_proba
    
    print("Random Forest trained!")
    print(f"AUC-ROC: {roc_auc_score(y_test, rf_pred_proba):.4f}")
    print(f"Accuracy: {(rf_pred == y_test).mean():.4f}")
    
    # Model 3: XGBoost
    print("\nTraining XGBoost...")
    xgb_model = xgb.XGBClassifier(
        n_estimators=100,
        max_depth=6,
        learning_rate=0.1,
        random_state=42,
        scale_pos_weight=(y_train == 0).sum() / (y_train == 1).sum(),
        eval_metric='auc'
    )
    xgb_model.fit(X_train, y_train)
    xgb_pred = xgb_model.predict(X_test)
    xgb_pred_proba = xgb_model.predict_proba(X_test)[:, 1]
    
    models['XGBoost'] = xgb_model
    predictions['XGBoost'] = xgb_pred
    predictions_proba['XGBoost'] = xgb_pred_proba
    
    print("XGBoost trained!")
    print(f"AUC-ROC: {roc_auc_score(y_test, xgb_pred_proba):.4f}")
    print(f"Accuracy: {(xgb_pred == y_test).mean():.4f}")
    
    # Model 4: LightGBM
    print("\nTraining LightGBM...")
    lgb_model = lgb.LGBMClassifier(
        n_estimators=100,
        max_depth=6,
        learning_rate=0.1,
        random_state=42,
        class_weight='balanced',
        verbose=-1
    )
    lgb_model.fit(X_train, y_train)
    lgb_pred = lgb_model.predict(X_test)
    lgb_pred_proba = lgb_model.predict_proba(X_test)[:, 1]
    
    models['LightGBM'] = lgb_model
    predictions['LightGBM'] = lgb_pred
    predictions_proba['LightGBM'] = lgb_pred_proba
    
    print("LightGBM trained!")
    print(f"AUC-ROC: {roc_auc_score(y_test, lgb_pred_proba):.4f}")
    print(f"Accuracy: {(lgb_pred == y_test).mean():.4f}")
    
    return models, predictions, predictions_proba, scaler

def evaluate_models(y_test, predictions, predictions_proba, output_dir):
    """Evaluate and compare models."""
    print("\n" + "="*80)
    print("MODEL EVALUATION")
    print("="*80)
    
    # ROC Curve
    plt.figure(figsize=(10, 8))
    for name, pred_proba in predictions_proba.items():
        fpr, tpr, _ = roc_curve(y_test, pred_proba)
        auc = roc_auc_score(y_test, pred_proba)
        plt.plot(fpr, tpr, label=f'{name} (AUC = {auc:.4f})', linewidth=2)
    
    plt.plot([0, 1], [0, 1], 'k--', label='Random', linewidth=1)
    plt.xlabel('False Positive Rate', fontsize=12)
    plt.ylabel('True Positive Rate', fontsize=12)
    plt.title('ROC Curve Comparison', fontsize=14, fontweight='bold')
    plt.legend(fontsize=10)
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig(f'{output_dir}/roc_curves.png', dpi=300, bbox_inches='tight')
    plt.close()
    print(f"\nROC curves saved to {output_dir}/roc_curves.png")
    
    # Model comparison table
    comparison = []
    for name, (pred, pred_proba) in zip(predictions.keys(), zip(predictions.values(), predictions_proba.values())):
        auc = roc_auc_score(y_test, pred_proba)
        accuracy = (pred == y_test).mean()
        comparison.append({
            'Model': name,
            'AUC-ROC': auc,
            'Accuracy': accuracy
        })
    
    comparison_df = pd.DataFrame(comparison)
    print("\nModel Comparison:")
    print("="*80)
    print(comparison_df.sort_values('AUC-ROC', ascending=False).to_string(index=False))
    
    return comparison_df

def feature_importance_analysis(models, features, X_train, output_dir):
    """Analyze feature importance."""
    print("\n" + "="*80)
    print("FEATURE IMPORTANCE ANALYSIS")
    print("="*80)
    
    # Feature importance (using Random Forest and XGBoost)
    fig, axes = plt.subplots(1, 2, figsize=(16, 6))
    
    # Random Forest feature importance
    if 'Random Forest' in models:
        rf_importance = pd.DataFrame({
            'feature': features,
            'importance': models['Random Forest'].feature_importances_
        }).sort_values('importance', ascending=False).head(20)
        
        axes[0].barh(range(len(rf_importance)), rf_importance['importance'])
        axes[0].set_yticks(range(len(rf_importance)))
        axes[0].set_yticklabels(rf_importance['feature'])
        axes[0].set_xlabel('Importance')
        axes[0].set_title('Random Forest - Top 20 Feature Importance')
        axes[0].invert_yaxis()
    
    # XGBoost feature importance
    if 'XGBoost' in models:
        xgb_importance = pd.DataFrame({
            'feature': features,
            'importance': models['XGBoost'].feature_importances_
        }).sort_values('importance', ascending=False).head(20)
        
        axes[1].barh(range(len(xgb_importance)), xgb_importance['importance'])
        axes[1].set_yticks(range(len(xgb_importance)))
        axes[1].set_yticklabels(xgb_importance['feature'])
        axes[1].set_xlabel('Importance')
        axes[1].set_title('XGBoost - Top 20 Feature Importance')
        axes[1].invert_yaxis()
    
    plt.tight_layout()
    plt.savefig(f'{output_dir}/feature_importance.png', dpi=300, bbox_inches='tight')
    plt.close()
    print(f"\nFeature importance plot saved to {output_dir}/feature_importance.png")
    
    if 'Random Forest' in models:
        print("\nTop 10 Features (Random Forest):")
        print(rf_importance.head(10).to_string(index=False))
    if 'XGBoost' in models:
        print("\nTop 10 Features (XGBoost):")
        print(xgb_importance.head(10).to_string(index=False))

def save_best_model(models, predictions_proba, y_test, scaler, output_dir):
    """Save the best performing model."""
    print("\n" + "="*80)
    print("SAVING BEST MODEL")
    print("="*80)
    
    # Find best model based on AUC-ROC
    best_model_name = None
    best_auc = 0
    
    for name, pred_proba in predictions_proba.items():
        auc = roc_auc_score(y_test, pred_proba)
        if auc > best_auc:
            best_auc = auc
            best_model_name = name
    
    if best_model_name:
        best_model = models[best_model_name]
        model_dir = Path(output_dir).parent / 'models'
        model_dir.mkdir(parents=True, exist_ok=True)
        
        # Save model and scaler
        joblib.dump(best_model, model_dir / 'best_model.pkl')
        joblib.dump(scaler, model_dir / 'scaler.pkl')
        
        print(f"\nBest Model: {best_model_name}")
        print(f"Best AUC-ROC: {best_auc:.4f}")
        print(f"Model saved to {model_dir / 'best_model.pkl'}")
        print(f"Scaler saved to {model_dir / 'scaler.pkl'}")
    else:
        print("No model to save!")

def main():
    """Main function to run ML analysis."""
    # Set up paths
    project_root = Path(__file__).parent.parent.parent
    data_path = project_root / 'data' / 'fraud_data.csv'
    output_dir = project_root / 'outputs' / 'figures'
    output_dir.mkdir(parents=True, exist_ok=True)
    
    # Load and prepare data
    X, y, features = load_and_prepare_data(data_path)
    
    # Split data into train and test sets
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42, stratify=y
    )
    
    print(f"\nTrain set: {X_train.shape}, Fraud rate: {y_train.mean():.4f}")
    print(f"Test set: {X_test.shape}, Fraud rate: {y_test.mean():.4f}")
    
    # Train models
    models, predictions, predictions_proba, scaler = train_models(
        X_train, X_test, y_train, y_test, features, output_dir
    )
    
    # Evaluate models
    comparison_df = evaluate_models(y_test, predictions, predictions_proba, output_dir)
    
    # Feature importance analysis
    feature_importance_analysis(models, features, X_train, output_dir)
    
    # Save best model
    save_best_model(models, predictions_proba, y_test, scaler, output_dir)
    
    # Summary
    print("\n" + "="*80)
    print("MACHINE LEARNING ANALYSIS SUMMARY")
    print("="*80)
    best_model_name = comparison_df.loc[comparison_df['AUC-ROC'].idxmax(), 'Model']
    best_auc = comparison_df['AUC-ROC'].max()
    print(f"\nBest Model: {best_model_name}")
    print(f"Best AUC-ROC: {best_auc:.4f}")
    print(f"\nModels trained: {len(models)}")
    print(f"Features used: {len(features)}")
    print("\nAnalysis complete!")
    print("="*80)

if __name__ == "__main__":
    main()

